docs/overview.rst

```diff
@@ -1,11 +1,410 @@
 Overview
 ========
 
-The data pipeline abstraction that Fuel provides looks complicated at first
-sight, but can be very powerful.
+We'll go over a quick example to see what Fuel is capable of.
+
+Let's start by creating some random data to act as features and targets. We'll
+pretend that we have eight 2x2 grayscale images separated into four classes.
+
+>>> import numpy
+>>> seed = 1234
+>>> rng = numpy.random.RandomState(seed)
+>>> features = rng.randint(256, size=(8, 2, 2))
+>>> targets = rng.randint(4, size=(8, 1))
+
+Our goal is to use Fuel to interface with this data, iterate over it in various
+ways and apply transformations to it on the fly.
+
+Division of labour
+------------------
+
+There are four basic tasks that Fuel needs to handle:
+
+* Interface with the data, be it on disk or in memory.
+* Decide which data points to visit, and in which order.
+* Iterate over the selected data points.
+* At each iteration step, apply some transformation to the selected data points.
+
+Each of those four tasks is delegated to a particular class of objects, which
+we'll be introducing in order.
+
+Datasets: interfacing with data
+-------------------------------
+
+.. admonition:: In summary
+    :class: tip
```

Comment:
I suggest a summary like this:

* **:class:`Dataset`**
    * It is an abstract class. Its subclasses are responsible for interfacing with your data.
    * **Constructor's arguments:**
        * ``sources``: optional, to be set to use only a subset of a dataset's sources.
    * **Instances' attributes:**
        - ``sources``: tuple of strings indicating which sources are provided by the dataset, and their ordering (which determines the return order of :meth:`get_data`).
        - ``axis_labels``: :class:`dict` mapping from source names to tuples of strings or ``None``. Used to document the axis semantics of the dataset's sources. Can be set via the ``axis_labels`` constructor argument.
        - ``num_examples``: when available, represents the number of examples the dataset holds.
    * **Instances' methods:**
        - :meth:`open`: returns a ``state`` object the dataset will interact with, e.g. a file handle, or ``None`` if stateless.
        - :meth:`get_data`: given a ``state`` object, returns data.
        - :meth:`close`: given a ``state`` object, properly closes it. 
        - :meth:`reset`: given a ``state`` object, properly closes it and returns a fresh one.

* **:class:`IterableDataset`**
    * Allows to interface with iterable objects in general.
    * The state returned by :meth:`IterableDataset.open` is an iterator object.
    * Can only iterate examplewise and sequentially.
    * **Constructor's arguments:**
        - ``iterables``: a :class:`dict` mapping from source names to their corresponding iterable objects.

* **:class:`IndexableDataset`**
    * Allows to interface with indexable objects in general.
    * Allows random data access.
    * Stateless, i.e., the state returned by :meth:`IndexableDataset.open` is ``None``.
    * **Constructor's arguments:**
        - ``indexables``: is expected to be a :class:`dict` mapping from source names to their corresponding indexable objects.
    * **Instances' methods:**
        - :meth:`get_data`: expects a list of indices as its ``request`` keyword argument.